src/Compilers/CSharp/Portable/Syntax/SyntaxFacts.cs

```diff
@@ -425,5 +426,50 @@ internal static bool IsDeclarationExpressionType(SyntaxNode node, out Declaratio
             parent = node.Parent as DeclarationExpressionSyntax;
             return node == parent?.Type;
         }
+
+        /// <summary>
+        /// Given an initializer expression infer the name of anonymous property.
+        /// Returns default(SyntaxToken) if unsuccessful
+        /// </summary>
+        public static SyntaxToken ExtractAnonymousTypeMemberName(this ExpressionSyntax input)
+        {
+            while (true)
+            {
+                switch (input.Kind())
+                {
+                    case SyntaxKind.IdentifierName:
+                        return ((IdentifierNameSyntax)input).Identifier;
+
+                    case SyntaxKind.SimpleMemberAccessExpression:
+                        input = ((MemberAccessExpressionSyntax)input).Name;
+                        continue;
+
+                    case SyntaxKind.ConditionalAccessExpression:
+                        input = ((ConditionalAccessExpressionSyntax)input).WhenNotNull;
+                        if (input.Kind() == SyntaxKind.MemberBindingExpression)
+                        {
+                            return ((MemberBindingExpressionSyntax)input).Name.Identifier;
+                        }
+
+                        continue;
+
+                    default:
+                        return default(SyntaxToken);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Checks whether the field name is reserved and tells us which position it's reserved for.
+        ///
+        /// For example:
+        /// Returns 3 for "Item3".
+        /// Returns 0 for "Rest", "ToString" and other members of System.ValueTuple.
+        /// Returns -1 for names that aren't reserved.
+        /// </summary>
+        public static int IsTupleElementNameReserved(string elementName)
```

Comment:
maybe "IsReservedTupleElementName"?   Though an 'is' method that is returning an int... 

maybe "GetTupleMemberNameIndex"?

Also, do we really give '3' for Item3?  Is this a 1-based API?